Setup e2e_node to support testing on ARM64

ike-ma opened the pull request:
* Build and create test artifacts for ARM64 using KUBE_BUILD_FOR_ARM64=true
* Prepull multi-arch ready container image
* Download ARM64 binaries/packages if running on ARM64 machine

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR enables the e2e_node to build for ARM64 and run on ARM64 machines

See https://github.com/kubernetes/test-infra/pull/29192  for the node_e2e setup for ARM64

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/test-infra/issues/28888

#### Special notes for your reviewer:

Successfully tested e2e_node locally on GCP t2a (arm64) machine family 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ike-ma committed:
Setup e2e_node to support testing on ARM64

* Enable dockerized build with --use-dockerized-build=true
* Build and create test artifacts for ARM64 with --target-build-arch=arm64
* Prepull multi-arch ready container image
* Download ARM64 binaries/packages if running on ARM64 machine

ike-ma commented:
/assign @bobbypage 
/cc @SergeyKanzhelev

ike-ma commented:
/cc @bobbypage

ike-ma commented on test/e2e_node/builder/build.go:
```diff
@@ -47,6 +47,10 @@ func BuildGo() error {
 	cmd := exec.Command("make", "-C", k8sRoot, fmt.Sprintf("WHAT=%s", targets))
+	if utils.IsBuildForArm64() {
+		klog.Infof("Building k8s binaries for Arm64: %s", targets)
+		cmd = exec.Command("make", "-C", k8sRoot, fmt.Sprintf("WHAT=%s", targets), "KUBE_BUILD_PLATFORMS=linux/arm64")
```
Note: this assumes `C compiler "aarch64-linux-gnu-gcc" ` is available on the build machine. 

Can be installed with `sudo apt-get install gcc-aarch64-linux-gnu`

ike-ma replied:
Thanks for the pointer. Added this to kubekins Dockerfile: https://github.com/kubernetes/test-infra/pull/29197

BenTheElder replied:
you already need to run docker right? you can use the dockerized make build here instead

BenTheElder replied:
looks like no, the CI job does not use docker

bobbypage replied:
Thinking about this some more, it seems like it would be beneficial to switch to use dockerized make build here. The benefit is that  dockerized make build already uses the kubecross image which supports building for arm (and has the all the needed cross build dependencies already installed). 

I would suggest the following:

1. Let's add a flag to the node e2e runner which controls if we use the dockerized build (e.g. `--use-dockerized-build`).
2. If the flag is enabled, build the artifacts here with ` build/run.sh make WHAT=cmd/kubelet KUBE_BUILD_PLATFORMS=linux/arm64` (+ the other targets needed).
3. In the test job for ARM, we enable  `--use-dockerized-build` and also set `privileged: true` since it requires nested docker

This also means we don't need to modify the kubekins image.

Long term, ideally we can switch all the jobs over to use dockerized build, it would be ideal as we already have some issues with issues like glibc incompatibility, but it will take some time to update all the jobs.

BenTheElder replied:
I think this should be a pre-req. We do not want to duplicate maintaining the build environment, and the kubekins image is already _massive_

ike-ma replied:
Done. Updated the current implementation with `--use-dockerized-build` flag and enable dockerized build instead of kubekins image.

BenTheElder commented on test/e2e_node/remote/utils.go:
```diff
@@ -60,14 +59,26 @@ providers:
+	if utils.IsBuildForArm64() {
+		cniArch = "arm64"
+	}
+	cniURL := "https://storage.googleapis.com/k8s-artifacts-cni/release/" + cniVersion + "/" + "cni-plugins-linux-" + cniArch + "-" + cniVersion + ".tgz"
```
what is this bucket? who is hosting this?

BenTheElder replied:
found it: https://github.com/kubernetes/kubernetes/pull/78819

bart0sh commented:
/triage accepted
/priority important-longterm

ike-ma commented on test/utils/arch.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+// IsRunningOnArm64 returns if tests is currently running on an Arm64 machine
+func IsRunningOnArm64() bool {
```
Done. Updated.

ike-ma commented on test/e2e_node/remote/utils.go:
```diff
@@ -60,14 +59,26 @@ providers:
     - "*.pkg.dev"
     defaultCacheDuration: 1m`
 
+func getCniURL() string {
```
Done. Updated for all

ike-ma commented:
Addressed all previous review comments.

Added a new flag `--target-build-arch=arm64` instead of using environment variable, tested with the following command, and works as expected

```
go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --target-build-arch=arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml
```

ike-ma commented:
Test Command

```
go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --target-build-arch=arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]|\[NodeFeature:DevicePluginProbe\]|\[NodeConformance\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml 
```

Test result
```
Ran 33 of 408 Specs in 1223.233 seconds
SUCCESS! -- 33 Passed | 0 Failed | 0 Pending | 375 Skipped
PASS
Ginkgo ran 1 suite in 20m23.445147507s
Test Suite Passed
```

logicalhan commented:
/remove-sig instrumentation

ike-ma commented:
/ ?

SergeyKanzhelev commented:
/assign @mmiranda96

mmiranda96 commented on test/e2e_node/remote/utils.go:
```diff
@@ -60,14 +59,25 @@ providers:
+	if builder.IsBuildForArm64() {
+		cniArch = "arm64"
+	}
+	cniURL := "https://storage.googleapis.com/k8s-artifacts-cni/release/" + cniVersion + "/" + "cni-plugins-linux-" + cniArch + "-" + cniVersion + ".tgz"
```
nit: replace string concat by `fmt.Sprintf()`.

ike-ma replied:
Done.

mmiranda96 commented on test/e2e_node/image_list.go:
```diff
@@ -73,6 +73,11 @@ var NodePrePullImageList = sets.NewString(
 func updateImageAllowList(ctx context.Context) {
+	// Architecture-specific image
+	if !utils.IsRunningOnArm64() {
+		//NodePerfTfWideDeep is only supported on x86_64, pulling in arm64 will fail
```
```suggestion
		// NodePerfTfWideDeep is only supported on x86_64, pulling in arm64 will fail
```

ike-ma replied:
Updated.

mmiranda96 reviewed (approved):
Just some small suggestions.

/gtm

mmiranda96 commented:
/ok-to-test

mmiranda96 commented:
/lgtm

bart0sh commented:
/retest

ike-ma commented:
Addressed all review comments.

Added ` --use-dockerized-build=true --target-build-arch=arm64 ` to trigger dockerized build for ARM64 test artifacts. 

# Test command

Tested locally with the following command

```
go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --use-dockerized-build=true --target-build-arch=arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]|\[NodeFeature:DevicePluginProbe\]|\[NodeConformance\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml 
```

# Revelant logs

```
I0517 00:22:06.076480  711172 remote.go:72] Building archive...
I0517 00:22:06.076612  711172 build.go:45] Building k8s binaries...
I0517 00:22:06.076741  711172 build.go:53] Building k8s binaries for Arm64: cmd/kubelet test/e2e_node/e2e_node.test github.com/onsi/ginkgo/v2/ginkgo cluster/gce/gci/mounter test/e2e_node/plugins/gcp-credential-provider
[0517 00:22:06] Verifying Prerequisites....
+++ [0517 00:22:06] Building Docker image kube-build:build-47e7333118-5-v1.27.0-go1.20.2-bullseye.0
+++ [0517 00:22:33] Syncing sources to container
+++ [0517 00:22:41] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [0517 00:22:41] Running build command...
Go version: go version go1.20.2 linux/amd64
+++ [0517 00:22:43] WARNING: linux/arm will no longer be built/shipped by default, please build it explicitly if needed.
+++ [0517 00:22:43]          support for linux/arm will be removed in a subsequent release.
+++ [0517 00:22:43] Building go targets for linux/arm64
    k8s.io/kubernetes/cmd/kubelet (non-static)
    k8s.io/kubernetes/test/e2e_node/e2e_node.test (test)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    k8s.io/kubernetes/cluster/gce/gci/mounter (non-static)
    k8s.io/kubernetes/test/e2e_node/plugins/gcp-credential-provider (non-static)
Env for linux/arm64: GOOS=linux GOARCH=arm64 GOROOT=/usr/local/go CGO_ENABLED=1 CC=aarch64-linux-gnu-gcc
Coverage is disabled.
+++ [0517 00:25:45] Placing binaries
+++ [0517 00:25:54] Syncing out of container
```

# Test Result

All serial tests works as expected on ARM64 machine

```
Ran 33 of 408 Specs in 1231.290 seconds
SUCCESS! -- 33 Passed | 0 Failed | 0 Pending | 375 Skipped
PASS

Ginkgo ran 1 suite in 20m31.490798452s
Test Suite Passed
Success Finished Test Suite on Host t2a-standard-2-ubuntu-gke-2204-1-24-arm64-v20230217-0ee04b54. Refer to artifacts directory for ginkgo log for this host.
```

bobbypage commented on test/e2e_node/builder/build.go:
```diff
@@ -47,6 +49,11 @@ func BuildGo() error {
 	}
 	targets := strings.Join(buildTargets, " ")
 	cmd := exec.Command("make", "-C", k8sRoot, fmt.Sprintf("WHAT=%s", targets))
+	if IsBuildForArm64() {
```
I think this should check `useDockerizedBuild` flag, not build for ARM. There are other usecases where dockerized build will be helpful and it shouldn't be restricted to just ARM.

I would suggest we use dockerized build if the dockerized flag is set. If ARM is set too as `target-build-arch` then, set `KUBE_BUILD_PLATFORMS`.

ike-ma replied:
I guess I can break this PR into two. In this PR, I want to highlight ARM only supports dockerized build. This one will focus on enable tests running on ARM machine, without the risk of introducing potential regression to x86 run. A follow-up PR will  enable x86 with useDockerizedBuild, and ensure there's no test regression with or without the `--use-dockerized-flag`

bobbypage replied:
I see... I would think if we set --use-dockerized-flag to false by default and use the old behavior if it's false, I think it should be safe: running the presubmit node e2e job would give good signal that nothing would be broken.

If the goal is to highlight ARM only supports dockerized build, maybe instead of defaulting, it would make sense to just return an error that specifies that for ARM, we need dockerized build. So something like:


```
if targetBuildArch == "arm64" && !isDockerizedBuild { 
    ... return error ...
}

if !isDockerizedBuild {
   ... do what we do today ...
} else {
   cmd = exec.Command(filepath.Join(k8sRoot, "build/run.sh"), "make", fmt.Sprintf("WHAT=%s", targets), "KUBE_BUILD_PLATFORMS=linux/" + targetBuildArch)
}
```

ike-ma replied:
Sounds good to me. I updated the semantics a bit for targetBuildArch (use 'linux/arm64' instead of 'arm64') so that `GetK8sBuildOutputDir()` doesn't need too much changes for KUBE_BUILD_PLATFORMS. I think now it should be pretty generic.

bobbypage commented on test/e2e_node/remote/node_e2e.go:
```diff
@@ -62,6 +62,10 @@ func (n *NodeE2ERemote) SetupTestPackage(tardir, systemSpecName string) error {
 	for _, bin := range requiredBins {
 		source := filepath.Join(buildOutputDir, bin)
+		if builder.IsBuildForArm64() {
+			source = filepath.Join(buildOutputDir, "linux_arm64", bin)
```
is this correct for dockerized mode? On my machine with dockerized build, the build directory ends up being, e.g. 

```
$ build/run.sh make WHAT=cmd/kubelet KUBE_BUILD_PLATFORMS=linux/arm64
$ ls _output/dockerized/bin/linux/arm64/kubelet
_output/dockerized/bin/linux/arm64/kubelet
```

Do we need to add also need to add  `/dockerized/` to the build output directory when it's building in dockerized mode?
I believe `GetK8sBuildOutputDir` assumes non-dockerized build... (xref: https://github.com/kubernetes/kubernetes/blob/master/test/utils/paths.go#L67-L78)

Additionally, we should also ensure this work for non ARM, i.e. dockerized build for amd64, so I don't think we should hardcode `linux_arm64`, but instead rely on the the target build arch flag.

ike-ma replied:
Hmm, I tested with this setup and the tests worked as expected: https://github.com/kubernetes/kubernetes/pull/117017#issuecomment-1550529226

Let me double check

ike-ma commented:
Support generic dockerized build  and be more arm64-agnostic.

Tested with `-use-dockerized-build=true --target-build-arch=linux/arm64 `


```
go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --use-dockerized-build=true --target-build-arch=linux/arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]|\[NodeFeature:DevicePluginProbe\]|\[NodeConformance\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml 
```

with relevant output as follow
```
I0518 01:32:37.507203 1210495 remote.go:72] Building archive...
I0518 01:32:37.507363 1210495 build.go:45] Building k8s binaries...
I0518 01:32:37.507476 1210495 build.go:53] Building dockerized k8s binaries targets cmd/kubelet test/e2e_node/e2e_node.test github.com/onsi/ginkgo/v2/ginkgo cluster/gce/gci/mounter test/e2e_node/plugins/gcp-credential-provider for architecture linux/arm64
+++ [0518 01:32:37] Verifying Prerequisites....
+++ [0518 01:32:38] Building Docker image kube-build:build-47e7333118-5-v1.27.0-go1.20.4-bullseye.0
+++ [0518 01:33:04] Syncing sources to container
+++ [0518 01:33:09] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [0518 01:33:09] Running build command...
+++ [0518 01:33:11] Setting GOMAXPROCS: 16
Go version: go version go1.20.4 linux/amd64
+++ [0518 01:33:11] Building go targets for linux/arm64
    k8s.io/kubernetes/cmd/kubelet (non-static)
    k8s.io/kubernetes/test/e2e_node/e2e_node.test (test)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    k8s.io/kubernetes/cluster/gce/gci/mounter (non-static)
    k8s.io/kubernetes/test/e2e_node/plugins/gcp-credential-provider (non-static)
Env for linux/arm64: GOOS=linux GOARCH=arm64 GOROOT=/usr/local/go CGO_ENABLED=1 CC=aarch64-linux-gnu-gcc
Coverage is disabled.
+++ [0518 01:36:49] Placing binaries
+++ [0518 01:36:57] Syncing out of container
I0518 01:37:06.324409 1210495 node_e2e.go:65] Copying binaries from /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/_output/dockerized/bin/linux/arm64/kubelet
I0518 01:37:06.433070 1210495 node_e2e.go:65] Copying binaries from /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/_output/dockerized/bin/linux/arm64/e2e_node.test
I0518 01:37:06.623577 1210495 node_e2e.go:65] Copying binaries from /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/_output/dockerized/bin/linux/arm64/ginkgo
I0518 01:37:06.648803 1210495 node_e2e.go:65] Copying binaries from /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/_output/dockerized/bin/linux/arm64/mounter
I0518 01:37:06.655077 1210495 node_e2e.go:65] Copying binaries from /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/_output/dockerized/bin/linux/arm64/gcp-credential-provider
```

jsturtevant commented:
/remove-sig windows
Windows doesn't run node e2e and don't currently have arm64 support in Windows Server skus

bobbypage commented:
Looks good, thank you for all the updates! This will not only add support for ARM, but also make it easier to use dockerized build which will help resolve build issues some folks have had. 

/lgtm

bobbypage commented:
/assign @SergeyKanzhelev

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

Thank you!

BenTheElder commented on test/utils/paths.go:
```diff
@@ -64,12 +64,17 @@ func RootDir() (string, error) {
-func GetK8sBuildOutputDir() (string, error) {
+// For dockerized build, targetArch (eg: 'linux/arm64', 'linux/amd64') must be explicitly specified
+// For non dockerized build, targetArch is ignored
+func GetK8sBuildOutputDir(isDockerizedBuild bool, targetArch string) (string, error) {
```
see also: https://github.com/kubernetes/kubernetes/blob/b2a1855f2c2241bab6ac21e4b1849f74237421af/hack/lib/util.sh#L205

I would recommend either sourcing this script or replicating the logic.

ike-ma replied:
Ack, I can potentially follow up in a separate PR. In the bash util, it looks for specific binary, and it checks which locations contains the latest binary build. Whereas for `GetK8sBuildOutputDir()`, it just returns a common build output dir, and rely on the client to check for whether the binary exists.

BenTheElder replied:
we'll want to minimize skew between how we select binaries for testing, this is what cluster e2e's use typically, we should at least match the search paths. TODO for followup

ike-ma replied:
Okay, I can follow up with a different PR for a quick refactor, I think the API would be something like `GetK8sBinary(binary, platform String)` instead of `GetK8sOutputDir`

BenTheElder commented on test/utils/arch.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+// IsRunningOnArm64 returns if tests is currently running on an arm64 machine
+func IsRunningOnArm64() bool {
```
nit: We only call this in one place, do we actually need a new public util here?

ike-ma replied:
Fair enough, I will remove this public util for now.

BenTheElder commented on test/e2e_node/remote/utils.go:
```diff
@@ -60,14 +59,25 @@ providers:
 
+func getCNIURL() string {
+	cniArch := "amd64"
+	if builder.IsBuildForArm64() {
```
TODO: And when we're using existing builds?

Seems like what we want is the target node architecture.

ike-ma replied:
Correct, this checks whether the targetArch is "linux/arm64"

BenTheElder replied:
AFAICT this checks the architecture we're _Building_ for, but you should be able to run node_e2e without building?

BenTheElder replied:
(or else, is at least confusingly named if not only checking for builds)

ike-ma replied:
This is a parameter passed in by user to indicate the target node architecture, I will update the name to `IsTargetArchArm64`

BenTheElder commented on test/e2e_node/remote/utils.go:
```diff
@@ -21,16 +21,15 @@ import (
 	"path/filepath"
 
 	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/test/e2e_node/builder"
```
nit: group local repo imports in their own group

ike-ma replied:
Will regroup.

BenTheElder replied:
this is also non-blocking but hopefully can go with other changes

ike-ma replied:
Done.

BenTheElder commented on test/e2e_node/image_list.go:
```diff
@@ -70,6 +70,11 @@ var NodePrePullImageList = sets.NewString(
 // So this function needs to be called after the extra envs are applied.
 func updateImageAllowList(ctx context.Context) {
+	// Architecture-specific image
+	if !utils.IsRunningOnArm64() {
```
The e2e binary can be running against a remote host, right? This doesn't seem like the correct check.

ike-ma replied:
My current testing is with `run_remote` where I trigger the test from x86 machine to execute test on the ARM64 machine. I think for current test-infra, this is how the e2e_node would be executed.

```
go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --use-dockerized-build=true --target-build-arch=arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]|\[NodeFeature:DevicePluginProbe\]|\[NodeConformance\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml 
```

I'm not too familiar with the case of executing e2e binary against a remote host. Do you have any suggestion on what's the most effective check for that case ?

BenTheElder replied:
that seems right, but then this logic is confusing, because it's the local process architecture.

I had to reread the remote runner to understand the e2e binary actually executes on the remote host.

for cluster e2es we usually just talk to the remote machine remotely and we've had broken assumptions versus the remote machines before, I over pattern matched on that :-)

BenTheElder commented on test/e2e_node/builder/build.go:
```diff
@@ -29,6 +29,8 @@ import (
 )
 
 var k8sBinDir = flag.String("k8s-bin-dir", "", "Directory containing k8s kubelet binaries.")
+var useDockerizedBuild = flag.Bool("use-dockerized-build", false, "Use dockerized build for test artifacts")
```
FWIW I recommend this as the default in the future for consistency reasons. You need docker available regardless to build images so switching doesn't change requirements.

ike-ma replied:
Ack. I can follow up with a separate PR to default dockerized build with more comphrehensive testing.

BenTheElder reviewed (commented):
Generally speaking it seems like we're using multiple different signals for which architecture is in play and none of them are actually coupled to the target node architecture, they're all the current process or the build output.

Build output seems fine, until we want to use an existing binary build instead of source building ...

Local process architecture just seems wrong, since we commonly execute e2e against remote hosts?

BenTheElder commented:
Generally LGTM, couple outstanding comments / clarifications

ike-ma commented:
Addressed Ben's comment, retested with `go run /usr/local/google/home/ikema/go-k8s-oss/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup -vmodule=*=4 --ssh-env=gce --results-dir=/tmp/e2e-node-results/ason --project=ikema-gke-dev-2 --use-dockerized-build=true --target-build-arch=linux/arm64 --zone=us-central1-a --ssh-user=ikema --ssh-key=/usr/local/google/home/ikema/.ssh/google_compute_engine --ginkgo-flags='--nodes=1 --focus="\[Serial\]" --skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeSpecialFeature\]|\[NodeAlphaFeature:.+\]|\[NodeAlphaFeature\]|\[NodeFeature:Eviction\]|\[NodeFeature:NodeProblemDetector\]|\[NodeFeature:OOMScoreAdj\]|\[NodeFeature:DevicePluginProbe\]|\[NodeConformance\]" '  --test-timeout=5h0m0s --image-config-file=/tmp/e2e-node-results/ason/image-config.yaml `

where I run the command locally on my x86 machine, and execute the test on Arm64 machine.

ike-ma commented:
/retest

BenTheElder reviewed (commented):
/lgtm
/approve
